src/Package/Impl/DataInspect/REnvironmentProvider.cs

```diff
@@ -44,41 +51,55 @@ internal class REnvironmentProvider : BindableBase, IREnvironmentProvider {
             }
         }
 
-        public async Task RefreshEnvironmentsAsync() {
-            await TaskUtilities.SwitchToBackgroundThread();
-
-            var envs = new List<REnvironment>();
-            bool success = false;
-            try {
-                var traceback = (await _rSession.TracebackAsync())
-                    .Skip(1) // skip global - it's in the search path already
-                    .Select(frame => new REnvironment(frame))
-                    .Reverse();
-                if (traceback.Any()) {
-                    envs.AddRange(traceback);
-                    envs.Add(null);
-                }
+        public async Task RefreshEnvironmentsAsync(CancellationToken cancellationToken = default(CancellationToken)) {
+            using (_disposeToken.Link(ref cancellationToken)) {
+                var session = _rSession;
 
-                var searchPath = (await _rSession.EvaluateAsync<string[]>("as.list(search())", REvaluationKind.BaseEnv))
-                    .Except(_hiddenEnvironments)
-                    .Select(name => new REnvironment(name));
-                envs.AddRange(searchPath);
+                await TaskUtilities.SwitchToBackgroundThread();
 
-                success = true;
-            } catch (RException) {
-            } catch (OperationCanceledException) {
-            }
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var envs = new List<REnvironment>();
+                bool success = false;
+                try {
+                    var traceback = (await session.TracebackAsync())
+                        .Skip(1) // skip global - it's in the search path already
+                        .Select(frame => new REnvironment(frame))
+                        .Reverse();
+                    if (traceback.Any()) {
+                        envs.AddRange(traceback);
+                        envs.Add(null);
+                    }
+
+                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I'd rather pass `cancellationToken` to `TracebackAsync` and `EvaluateAsync` instead of calling `ThrowIfCancellationRequested` between those calls